File: src/fs_explorer/explore_sessions.py
"""
In-memory session management for SSE-based exploration workflows.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any
from uuid import uuid4


def utc_now() -> datetime:
    """Return the current UTC timestamp."""
    return datetime.now(timezone.utc)


def isoformat_utc(value: datetime) -> str:
    """Serialize timestamps in a stable UTC format."""
    return value.astimezone(timezone.utc).isoformat()


@dataclass(slots=True)
class ExploreStreamEvent:
    """A single event emitted to SSE subscribers."""

    sequence: int
    type: str
    timestamp: datetime
    data: dict[str, Any]

    def as_payload(self, session_id: str) -> dict[str, Any]:
        """Convert the event to the wire payload."""
        return {
            "session_id": session_id,
            "type": self.type,
            "sequence": self.sequence,
            "timestamp": isoformat_utc(self.timestamp),
            "data": self.data,
        }


@dataclass(slots=True)
class ExploreSession:
    """Runtime state for one exploration run."""

    session_id: str
    task: str
    document_ids: list[str]
    collection_id: str | None
    db_path: str | None
    enable_semantic: bool
    enable_metadata: bool
    status: str = "created"
    pending_question: str | None = None
    final_result: str | None = None
    error: str | None = None
    last_focus_anchor: dict[str, Any] | None = None
    context_budget_stats: dict[str, Any] = field(default_factory=dict)
    context_state_snapshot: dict[str, Any] = field(default_factory=dict)
    lazy_indexing_stats: dict[str, Any] = field(
        default_factory=lambda: {
            "triggered": False,
            "indexed_documents": 0,
            "chunks_written": 0,
            "embeddings_written": 0,
        }
    )
    created_at: datetime = field(default_factory=utc_now)
    updated_at: datetime = field(default_factory=utc_now)
    history: list[ExploreStreamEvent] = field(default_factory=list)
    subscribers: set[asyncio.Queue[ExploreStreamEvent | None]] = field(default_factory=set)
    handler: Any | None = None
    workflow_task: asyncio.Task[None] | None = None
    _next_sequence: int = 1
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock)

    def snapshot(self) -> dict[str, Any]:
        """Return a JSON-friendly snapshot of the session state."""
        return {
            "session_id": self.session_id,
            "status": self.status,
            "awaiting_human": self.status == "awaiting_human",
            "pending_question": self.pending_question,
            "final_result": self.final_result,
            "error": self.error,
            "last_focus_anchor": self.last_focus_anchor,
            "context_budget_stats": self.context_budget_stats,
            "context_state_snapshot": self.context_state_snapshot,
            "lazy_indexing_stats": self.lazy_indexing_stats,
            "task": self.task,
            "document_ids": list(self.document_ids),
            "collection_id": self.collection_id,
            "db_path": self.db_path,
            "enable_semantic": self.enable_semantic,
            "enable_metadata": self.enable_metadata,
            "created_at": isoformat_utc(self.created_at),
            "updated_at": isoformat_utc(self.updated_at),
        }

    def is_terminal(self) -> bool:
        """Return whether the session is finished."""
        return self.status in {"completed", "error", "closed"}

    async def publish(self, event_type: str, data: dict[str, Any]) -> ExploreStreamEvent:
        """Record a new event and fan it out to current subscribers."""
        async with self._lock:
            event = ExploreStreamEvent(
                sequence=self._next_sequence,
                type=event_type,
                timestamp=utc_now(),
                data=data,
            )
            self._next_sequence += 1
            self.updated_at = event.timestamp
            self.history.append(event)
            subscribers = list(self.subscribers)

        for subscriber in subscribers:
            subscriber.put_nowait(event)
        return event


class ExploreSessionManager:
    """Registry for in-flight and recently completed exploration sessions."""

    def __init__(self, *, retention_minutes: int = 15) -> None:
        self._sessions: dict[str, ExploreSession] = {}
        self._retention = timedelta(minutes=retention_minutes)
        self._lock = asyncio.Lock()

    async def create_session(
        self,
        *,
        task: str,
        document_ids: list[str],
        collection_id: str | None,
        db_path: str | None,
        enable_semantic: bool,
        enable_metadata: bool,
    ) -> ExploreSession:
        """Create and register a new session."""
        await self.cleanup()
        session = ExploreSession(
            session_id=uuid4().hex,
            task=task,
            document_ids=list(document_ids),
            collection_id=collection_id,
            db_path=db_path,
            enable_semantic=enable_semantic,
            enable_metadata=enable_metadata,
        )
        async with self._lock:
            self._sessions[session.session_id] = session
        return session

    async def get_session(self, session_id: str) -> ExploreSession | None:
        """Fetch a session by id."""
        await self.cleanup()
        async with self._lock:
            return self._sessions.get(session_id)

    async def subscribe(
        self,
        session_id: str,
    ) -> tuple[
        ExploreSession | None,
        asyncio.Queue[ExploreStreamEvent | None] | None,
        list[ExploreStreamEvent] | None,
    ]:
        """Register a live subscriber queue for a session."""
        session = await self.get_session(session_id)
        if session is None:
            return None, None, None

        queue: asyncio.Queue[ExploreStreamEvent | None] = asyncio.Queue()
        async with session._lock:
            history = list(session.history)
            session.subscribers.add(queue)
        return session, queue, history

    async def unsubscribe(
        self,
        session: ExploreSession,
        queue: asyncio.Queue[ExploreStreamEvent | None],
    ) -> None:
        """Remove a live subscriber queue."""
        async with session._lock:
            session.subscribers.discard(queue)

    async def cleanup(self) -> None:
        """Drop expired terminal sessions."""
        cutoff = utc_now() - self._retention
        async with self._lock:
            expired = [
                session_id
                for session_id, session in self._sessions.items()
                if session.is_terminal() and session.updated_at < cutoff
            ]
            for session_id in expired:
                self._sessions.pop(session_id, None)


def encode_sse_event(session_id: str, event: ExploreStreamEvent) -> str:
    """Serialize an event using the SSE wire format."""
    payload = json.dumps(event.as_payload(session_id), ensure_ascii=False)
    return f"event: {event.type}\ndata: {payload}\n\n"

File: src/runtime/explore-sessions.ts
/*
Reference: legacy/python/src/fs_explorer/explore_sessions.py
Reference: legacy/python/src/fs_explorer/server.py
*/

import { randomUUID } from "node:crypto";

import type {
  BatchMode,
  ExploreSessionSnapshot,
  ExploreSessionStatus,
  ExploreStreamEventPayload,
} from "../types/exploration.js";

export function utcNow(): Date {
  return new Date();
}

export function isoformatUtc(value: Date): string {
  return value.toISOString();
}

export class ExploreStreamEvent {
  constructor(
    readonly sequence: number,
    readonly type: string,
    readonly timestamp: Date,
    readonly data: Record<string, unknown>,
  ) {}

  asPayload(sessionId: string): ExploreStreamEventPayload {
    return {
      session_id: sessionId,
      type: this.type,
      sequence: this.sequence,
      timestamp: isoformatUtc(this.timestamp),
      data: this.data,
    };
  }
}

export class AsyncEventQueue<T> implements AsyncIterable<T> {
  private readonly values: T[] = [];

  private readonly waiters: Array<(value: IteratorResult<T>) => void> = [];

  private closed = false;

  push(value: T): void {
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter({ value, done: false });
      return;
    }
    this.values.push(value);
  }

  close(): void {
    this.closed = true;
    for (const waiter of this.waiters.splice(0)) {
      waiter({ value: undefined as T, done: true });
    }
  }

  next(): Promise<IteratorResult<T>> {
    const value = this.values.shift();
    if (value !== undefined) {
      return Promise.resolve({ value, done: false });
    }
    if (this.closed) {
      return Promise.resolve({ value: undefined as T, done: true });
    }
    return new Promise((resolve) => {
      this.waiters.push(resolve);
    });
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return {
      next: () => this.next(),
    };
  }
}

export class ExploreSession {
  status: ExploreSessionStatus = "created";

  pendingQuestion: string | null = null;

  finalResult: string | null = null;

  error: string | null = null;

  lastFocusAnchor: Record<string, unknown> | null = null;

  contextBudgetStats: Record<string, unknown> = {};

  contextStateSnapshot: Record<string, unknown> = {};

  lazyIndexingStats: Record<string, unknown> = {
    triggered: false,
    indexed_documents: 0,
    chunks_written: 0,
    embeddings_written: 0,
  };

  candidateDocumentSelection: Record<string, unknown> | null = null;

  documentSummaries: Array<Record<string, unknown>> = [];

  parallelDocumentLimit = 3;

  batchSummaries: Array<Record<string, unknown>> = [];

  cumulativeAnswer: string | null = null;

  readonly createdAt = utcNow();

  updatedAt = this.createdAt;

  readonly history: ExploreStreamEvent[] = [];

  readonly subscribers = new Set<AsyncEventQueue<ExploreStreamEvent>>();

  workflowTask: Promise<void> | null = null;

  runtime: unknown = null;

  private nextSequence = 1;

  constructor(
    readonly sessionId: string,
    readonly task: string,
    readonly documentIds: string[],
    readonly collectionId: string | null,
    readonly collectionIds: string[],
    readonly dbPath: string | null,
    readonly enableSemantic: boolean,
    readonly enableMetadata: boolean,
    readonly batchMode: BatchMode,
    readonly batchSize: number,
    readonly batchThreshold: number,
  ) {}

  snapshot(): ExploreSessionSnapshot {
    return {
      session_id: this.sessionId,
      status: this.status,
      awaiting_human: this.status === "awaiting_human",
      pending_question: this.pendingQuestion,
      final_result: this.finalResult,
      error: this.error,
      last_focus_anchor: this.lastFocusAnchor,
      context_budget_stats: this.contextBudgetStats,
      context_state_snapshot: this.contextStateSnapshot,
      lazy_indexing_stats: this.lazyIndexingStats,
      task: this.task,
      document_ids: [...this.documentIds],
      collection_id: this.collectionId,
      collection_ids: [...this.collectionIds],
      db_path: this.dbPath,
      enable_semantic: this.enableSemantic,
      enable_metadata: this.enableMetadata,
      candidate_document_selection: this.candidateDocumentSelection,
      document_summaries: [...this.documentSummaries],
      parallel_document_limit: this.parallelDocumentLimit,
      batch_summaries: [...this.batchSummaries],
      cumulative_answer: this.cumulativeAnswer,
      batch_mode: this.batchMode,
      batch_size: this.batchSize,
      batch_threshold: this.batchThreshold,
      created_at: isoformatUtc(this.createdAt),
      updated_at: isoformatUtc(this.updatedAt),
    };
  }

  isTerminal(): boolean {
    return ["completed", "error", "closed"].includes(this.status);
  }

  publish(eventType: string, data: Record<string, unknown>): ExploreStreamEvent {
    const event = new ExploreStreamEvent(this.nextSequence, eventType, utcNow(), data);
    this.nextSequence += 1;
    this.updatedAt = event.timestamp;
    this.history.push(event);
    for (const subscriber of this.subscribers) {
      subscriber.push(event);
    }
    return event;
  }
}

export class ExploreSessionManager {
  private readonly sessions = new Map<string, ExploreSession>();

  private readonly retentionMs: number;

  constructor(input: { retentionMinutes?: number } = {}) {
    this.retentionMs = (input.retentionMinutes ?? 15) * 60 * 1000;
  }

  createSession(input: {
    task: string;
    documentIds: string[];
    collectionId?: string | null;
    collectionIds?: string[] | null;
    dbPath?: string | null;
    enableSemantic?: boolean;
    enableMetadata?: boolean;
    batchMode?: BatchMode;
    batchSize?: number | null;
    batchThreshold?: number | null;
  }): ExploreSession {
    this.cleanup();
    const session = new ExploreSession(
      randomUUID().replaceAll("-", ""),
      input.task,
      [...input.documentIds],
      input.collectionId ?? null,
      [
        ...new Set(
          [...(input.collectionIds ?? []), ...(input.collectionId ? [input.collectionId] : [])]
            .map((item) => String(item).trim())
            .filter(Boolean),
        ),
      ],
      input.dbPath ?? null,
      Boolean(input.enableSemantic),
      Boolean(input.enableMetadata),
      input.batchMode ?? "auto",
      Math.max(Number(input.batchSize ?? 5), 1),
      Math.max(Number(input.batchThreshold ?? 10), 1),
    );
    this.sessions.set(session.sessionId, session);
    return session;
  }

  getSession(sessionId: string): ExploreSession | null {
    this.cleanup();
    return this.sessions.get(sessionId) ?? null;
  }

  subscribe(sessionId: string): {
    session: ExploreSession | null;
    queue: AsyncEventQueue<ExploreStreamEvent> | null;
    history: ExploreStreamEvent[] | null;
  } {
    const session = this.getSession(sessionId);
    if (!session) {
      return { session: null, queue: null, history: null };
    }
    const queue = new AsyncEventQueue<ExploreStreamEvent>();
    session.subscribers.add(queue);
    return { session, queue, history: [...session.history] };
  }

  unsubscribe(session: ExploreSession, queue: AsyncEventQueue<ExploreStreamEvent>): void {
    session.subscribers.delete(queue);
    queue.close();
  }

  cleanup(): void {
    const cutoff = Date.now() - this.retentionMs;
    for (const [sessionId, session] of this.sessions) {
      if (session.isTerminal() && session.updatedAt.getTime() < cutoff) {
        this.sessions.delete(sessionId);
      }
    }
  }
}

export function encodeSseEvent(sessionId: string, event: ExploreStreamEvent): string {
  return `event: ${event.type}\ndata: ${JSON.stringify(event.asPayload(sessionId))}\n\n`;
}
